styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java

```diff
@@ -308,6 +320,39 @@ public void updateNextNaturalTrigger(WorkflowId workflowId, TriggerInstantSpec t
     return map;
   }
 
+  public Map<WorkflowId, Workflow> workflows(Set<WorkflowId> workflowIds) {
+    final Callable<Map<WorkflowId, Workflow>> callable =
```

Comment:
I'm a little uneasy with this as FJP and java streams are designed for parallel _computation_, not blocking IO. 

https://stackoverflow.com/questions/27584271/stream-processing-thread-pool-exhaustion#comment43608638_27584271

http://gee.cs.oswego.edu/dl/html/StreamParallelGuidance.html

I'm not sure how FJP will behave wrt blocked worker threads unless the `ManagedBlocker` is used.

https://zeroturnaround.com/rebellabs/fixedthreadpool-cachedthreadpool-or-forkjoinpool-picking-correct-java-executors-for-background-tasks/

Reply:
Yeah I read those articles. It seems the concern is about using the common pool with blocking IO. And that for sure will impact overall JVM performance. ManagedBlocker gives a hint to create more workers to make enlarge common pool if I understand it correctly.

Reply:
Did some simple experiments (on my 4 core laptop) with managed vs unmanaged blocking of dedicated FJP workers and the results surprised me.

When using parallel stream, the number of threads used would never go above 16, even when configuring the FJP concurrency to a high value like 1000. This remained true even when using `managedBlock`. However, doing many raw `submit()` calls produced the expected level of concurrency.

Setting the FJP concurrency to a number lower than 16 capped the effective number of worker threads used to that number.

Turns out that java parallel streams limit the concurrency, independently of the underlying FJP. 

https://stackoverflow.com/a/37518272

Reply:
https://stackoverflow.com/a/29272776

Reply:
I've verified that the parallel stream concurrency behavior changes in JDK 10. In JDK 10 the full dedicated FJP concurrency is utilized.

Reply:
I have started to question the actual benefits getting batches in parallel vs. complexity using a thread pool, future, wait and then merge. We only do 3-4 batches normally. Maybe something we can use in CompletableFurtures?

Reply:
Btw even with the buggy FJP, the performance reading workflows was improved around 30%. 4s vs. 6s